src/section_list_cache.rs

```diff
@@ -0,0 +1,159 @@
+use routing_table::Prefix;
+use rust_sodium::crypto::sign::Signature;
+use std::collections::HashMap;
+use super::QUORUM;
+use super::XorName;
+
+pub type Signatures = HashMap<PublicId, Signature>;
+pub type PrefixMap<T> = HashMap<Prefix<XorName>, T>;
+
+pub struct SectionListCache {
+    // all signatures for a group list for a given prefix
+    signatures: PrefixMap<HashMap<GroupList, Signatures>>,
+    // group lists signed by a given public id
+    signed_by: HashMap<PublicId, PrefixMap<GroupList>>,
+    // the latest group list for each prefix with a quorum of signatures
+    lists_cache: PrefixMap<(GroupList, Signatures)>,
+}
+
+impl SectionListCache {
+    pub fn new() -> SectionListCache {
+        SectionListCache {
+            signatures: Default::default(),
+            signed_by: Default::default(),
+            lists_cache: Default::default(),
+        }
+    }
+
+    fn prune(&mut self) {
+        let mut to_remove = vec![];
+        for (prefix, map) in &mut self.signatures {
+            // prune group lists with 0 signatures
+            let lists_to_remove = map.iter()
+                .filter(|&(_, sigs)| sigs.is_empty())
+                .map(|(list, _)| list.clone())
+                .collect_vec();
+            for list in lists_to_remove {
+                let _ = map.remove(&list);
+            }
+            if map.is_empty() {
+                to_remove.push(*prefix);
+            }
+        }
+        // prune prefixes with no group lists
+        for prefix in to_remove {
+            let _ = self.signatures.remove(&prefix);
+        }
+
+        let to_remove = self.signed_by
+            .iter()
+            .filter(|&(_, map)| map.is_empty())
+            .map(|(pub_id, _)| *pub_id)
+            .collect_vec();
+        // prune pub_ids signing nothing
+        for pub_id in to_remove {
+            let _ = self.signed_by.remove(&pub_id);
+        }
+    }
+
+    fn update_lists_cache(&mut self, our_section_size: usize) {
+        for (prefix, map) in &self.signatures {
+            // find the entries with the most signatures
+            let entries = map.iter()
+                .map(|(list, sigs)| (list, sigs.len()))
+                .sorted_by(|lhs, rhs| rhs.1.cmp(&lhs.1));
+            if let Some(&entry) = entries.first() {
+                // entry.0 = list, entry.1 = num of signatures
+                if 100 * entry.1 >= QUORUM * our_section_size {
+                    // we have a list with a quorum of signatures
+                    let signatures = unwrap!(map.get(entry.0));
+                    let _ = self.lists_cache.insert(*prefix, (entry.0.clone(), signatures.clone()));
+                }
+            }
+        }
+    }
+
+    fn remove_signatures_for_prefix_by(&mut self, prefix: Prefix<XorName>, author: PublicId) {
+        // vector of tuples (prefix, group list) to be removed
+        let to_remove = self.signed_by
+            .get(&author)
+            .map_or(Default::default(), |map| {
```

Comment:
Might make sense, true. `Default` leaves more room for changes when we decide that some other structure is more efficient, but it might not be worth it.